pkg/storage/storage.go

```diff
@@ -137,11 +137,33 @@ type ReadStartingWithUserOptions struct {
 	WithResultsSortedAscending bool
 }
 
+type OnDuplicate int32
+
+const (
+	OnDuplicateUnspecified OnDuplicate = 0 // If set to 'OnDuplicateUnspecified' OR not set, the API will behave as if it was set to 'OnDuplicateError'.
+	OnDuplicateError       OnDuplicate = 1 // Return an error if a tuple already exists.
+	OnDuplicateIgnore      OnDuplicate = 2 // Treat identical writes as no-ops if all attributes (including the RelationshipCondition) match.
+)
+
 // Writes is a typesafe alias for Write arguments.
-type Writes = []*openfgav1.TupleKey
+type Writes struct {
```

Comment:
Please be careful to call out this change in the changelog. A few adopters (e.g. canonical, godaddy, etc..) do not consume the public API and use the datastore interface directly